File: fieldmark-go/internal/domain/entities/audit_entry.go
package entities

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// AuditEntry mirrors domain.audit_entry. Write-once value object; behavior
// methods are intentionally absent (Story 2.2). The Action field is a plain
// string rather than the enums.AuditAction type so unrecognized DB values
// surface to the caller without enum coercion — Story 2.13's audit-log read
// path can decide how to surface the divergence.
//
// JSONB columns use json.RawMessage so the storage layer is opaque about
// payload shape. Nil means absence-of-payload (semantically distinct from
// `'null'::jsonb`); the helper must preserve the distinction.
type AuditEntry struct {
	ID          uuid.UUID
	OccurredAt  time.Time
	ActorID     uuid.UUID
	Action      string
	EntityType  string
	EntityID    uuid.UUID
	ProjectID   *uuid.UUID
	BeforeState json.RawMessage
	AfterState  json.RawMessage
	Metadata    json.RawMessage
}

File: FieldMark/FieldMark.Domain/Entities/AuditEntry.cs
using System.Text.Json;

namespace FieldMark.Domain.Entities;

// Write-once value object for domain.audit_entry. The persisted Action column
// is stored as a string rather than the AuditAction enum so an unrecognized DB
// value surfaces as a deserialization failure on read in a future audit-read
// path (Story 2.13) rather than silently corrupting the row.
//
// Helpers (IAuditAppender) live in FieldMark.Data; this type is pure data.
public sealed class AuditEntry
{
    public Guid Id { get; private set; }
    public DateTimeOffset OccurredAt { get; private set; }
    public Guid ActorId { get; private set; }
    public string Action { get; private set; } = string.Empty;
    public string EntityType { get; private set; } = string.Empty;
    public Guid EntityId { get; private set; }
    public Guid? ProjectId { get; private set; }
    public JsonDocument? BeforeState { get; private set; }
    public JsonDocument? AfterState { get; private set; }
    public JsonDocument? Metadata { get; private set; }

    private AuditEntry() { }

    public AuditEntry(
        Guid actorId,
        string action,
        string entityType,
        Guid entityId,
        Guid? projectId = null,
        JsonDocument? beforeState = null,
        JsonDocument? afterState = null,
        JsonDocument? metadata = null
    )
    {
        Id = Guid.NewGuid();
        ActorId = actorId;
        Action = action;
        EntityType = entityType;
        EntityId = entityId;
        ProjectId = projectId;
        BeforeState = beforeState;
        AfterState = afterState;
        Metadata = metadata;
    }
}
